File: ios/Packages/Bolsillo/Sources/BolsilloDomain/UseCases/RecordTransfer.swift
import Foundation

/// Records a transfer as two linked `Transaction` legs in a single atomic operation
/// (Constitution Article III — Invariant 6: transfers are always double-entry pairs).
///
/// Invariant enforcement:
/// - Source leg is negative, destination leg is positive (Invariant 1).
/// - Both legs share the same `transferGroupId` (Invariant 6).
/// - Cross-currency transfers are rejected in feature 001; same-currency only (E8 deferred).
/// - `fxRateMillis` is frozen to 1000 and `amountBase == amount` (Invariant 8).
public struct RecordTransfer: Sendable {
    private let transactions: any TransactionRepository
    private let accounts: any AccountRepository

    public init(transactions: some TransactionRepository, accounts: some AccountRepository) {
        self.transactions = transactions
        self.accounts = accounts
    }

    /// Atomically persist a transfer and return both legs.
    ///
    /// - Parameters:
    ///   - sourceAccountId: The account being debited.
    ///   - destAccountId: The account being credited.
    ///   - amount: The transfer magnitude (positive minor-unit value).
    ///   - occurredAt: The effective date of the transfer.
    /// - Throws: `TransferError.sameAccount` if both ids are identical.
    ///           `TransferError.crossCurrency` if the accounts use different currencies.
    public func callAsFunction(
        sourceAccountId: String,
        destAccountId: String,
        amount: Money,
        occurredAt: Date
    ) async throws -> TransferPair {
        guard sourceAccountId != destAccountId else { throw TransferError.sameAccount }

        // Resolve source currency (used for both legs in same-currency 001)
        let srcAccount = await accounts.getById(sourceAccountId)
        let dstAccount = await accounts.getById(destAccountId)

        if let src = srcAccount, let dst = dstAccount, src.currencyCode != dst.currencyCode {
            throw TransferError.crossCurrency
        }

        let currencyCode = srcAccount?.currencyCode ?? "USD"
        let groupId = UUID().uuidString
        let now = Date()

        let legSource = Transaction(
            id: UUID().uuidString,
            accountId: sourceAccountId,
            type: .transfer,
            amount: -amount,            // negative: source is debited (Invariant 1)
            currencyCode: currencyCode,
            amountBase: -amount,        // frozen at creation (Invariant 8)
            fxRateMillis: 1000,         // same-currency (Invariant 8)
            categoryId: "transfer",
            merchant: nil,
            note: nil,
            occurredAt: occurredAt,
            transferGroupId: groupId,
            createdAt: now,
            updatedAt: now
        )

        let legDest = Transaction(
            id: UUID().uuidString,
            accountId: destAccountId,
            type: .transfer,
            amount: amount,             // positive: destination is credited (Invariant 1)
            currencyCode: currencyCode,
            amountBase: amount,         // frozen at creation (Invariant 8)
            fxRateMillis: 1000,         // same-currency (Invariant 8)
            categoryId: "transfer",
            merchant: nil,
            note: nil,
            occurredAt: occurredAt,
            transferGroupId: groupId,
            createdAt: now,
            updatedAt: now
        )

        await transactions.upsertTransfer(legSource: legSource, legDest: legDest)
        return TransferPair(legSource: legSource, legDest: legDest)
    }
}

File: android/domain/bin/main/com/bolsillo/domain/usecase/RecordTransfer.kt
package com.bolsillo.domain.usecase

import com.bolsillo.domain.model.Money
import com.bolsillo.domain.model.Transaction
import com.bolsillo.domain.model.TransactionType
import com.bolsillo.domain.port.AccountRepository
import com.bolsillo.domain.port.TransactionRepository
import java.util.UUID
import javax.inject.Inject

data class TransferPair(val source: Transaction, val destination: Transaction)

/**
 * FR 9,10 — atomic linked double entry. Same-account rejected (FR 10),
 * cross-currency rejected (Invariant 6 — E8 will lift this). FX is frozen at
 * creation on both legs (Invariant 8 — same-currency 001 keeps rate = 1000).
 */
class RecordTransfer
    @Inject
    constructor(
        private val transactions: TransactionRepository,
        private val accounts: AccountRepository,
    ) {
        suspend operator fun invoke(
            sourceAccountId: String,
            destAccountId: String,
            amount: Money,
            occurredAt: Long,
            now: Long = System.currentTimeMillis(),
            groupIdOverride: String? = null,
        ): TransferPair {
            require(sourceAccountId != destAccountId) { "record.transfer.sameAccountError" }
            require(amount.minorUnits > 0L) { "Transfer amount must be a positive magnitude" }
            val src =
                accounts.getById(sourceAccountId)
                    ?: error("Source account not found: $sourceAccountId")
            val dest =
                accounts.getById(destAccountId)
                    ?: error("Destination account not found: $destAccountId")
            require(src.currencyCode == dest.currencyCode) {
                "Cross-currency transfers are not supported in spec 001"
            }
            val groupId = groupIdOverride ?: UUID.randomUUID().toString()
            val signedMinor = amount.minorUnits
            val legSource =
                Transaction(
                    id = UUID.randomUUID().toString(),
                    accountId = sourceAccountId,
                    type = TransactionType.TRANSFER,
                    amount = Money(-signedMinor),
                    currencyCode = src.currencyCode,
                    amountBase = Money(-signedMinor),
                    fxRateMillis = 1000L,
                    categoryId = "transfer",
                    occurredAt = occurredAt,
                    transferGroupId = groupId,
                    createdAt = now,
                    updatedAt = now,
                )
            val legDest =
                Transaction(
                    id = UUID.randomUUID().toString(),
                    accountId = destAccountId,
                    type = TransactionType.TRANSFER,
                    amount = Money(signedMinor),
                    currencyCode = dest.currencyCode,
                    amountBase = Money(signedMinor),
                    fxRateMillis = 1000L,
                    categoryId = "transfer",
                    occurredAt = occurredAt,
                    transferGroupId = groupId,
                    createdAt = now,
                    updatedAt = now,
                )
            transactions.upsertTransfer(legSource, legDest)
            return TransferPair(legSource, legDest)
        }
    }
